Add .NET Core build

kal opened the pull request:
This PR adds a project that builds a .NET Core version of MedallionShell and adds it to the main solution. Also included is a nuspec file to make it easy to package up the two assemblies into a single .nupkg.

Changes required are quite minor and are all contained in conditional sections (using the NETCORE compiler symbol).

I haven't bumped version numbers as I'll leave it to you to decide if this warrants a change to minor version or just to patch version - though I'd probably go with calling it 1.2.0 :-)

Commit:
Add .NET Core project

Targets netstandard1.3 as this is the lowest platform that is supported
by the System.Diagnostics.Process dependency.

Commit:
Add the .NET Core project to the solution

NOTE: This change also bumps the VS version up to VS 14

Commit:
Conditional compilation options for .NET Core support

No ReadAsync/WriteAsync and no Close method on streams.
No support for environment variables in ProcessStartInfo

Commit:
Add nuspec package

Enables easy creation of the NuGet package with both .NET 4.5 and .NET
Core libraries included.

Commit:
Ignore VS outputs

madelson commented:
@kal this is great. I will try to take a more in-depth look over the next few days. One question I have is how you tested this. Do the unit tests run under both configurations or do those need some work as well?

kal commented:
@madelson - I did some testing on a separate branch [feature-netcore](https://github.com/NetworkedPlanet/MedallionShell/tree/feature-netcore). I didn't include that in the PR as it involves migrating all of your tests over to xUnit and changing the SampleCommand project to drop its output into the test project directory. Also on that branch I made the mistake of changing your existing MedallionShell project into a .NET Core project rather than adding a second parallel project like I did in this PR.

I could look at how to replicate what I did on that branch without hacking your existing tests (maybe just be creating a duplicate project with cut-n-pasted code), if that helps?

madelson commented:
@kal if you've already gone through the work of converting to xUnit so that both builds can be tested that would be great to include in the PR as well. The only reason I'm using MSTest is because I wrote this pre VS2015 and I wanted a test project that anyone could run in VS with just the express edition. With VS2015 community I think you can run xUnit just fine so I'd be happy to switch over to that.

Commit:
XUnit Tests

Migrated .NET tests to XUnit and added .NET Core test project that
shares the same source.

kal commented:
@madelson I've pushed a commit that migrates the unit tests over to XUnit and has a separate test project for .NET Core. Due to some limitations in the scriptability of project.json files I've tweaked the .csproj of SampleCommand to copy the output .exe over to the MedallionShell.NetCore.Tests directory so that it can be used in the tests.

All the unit tests pass, except that if you run all the tests together then the test `Test Name:    Medallion.Shell.Tests.Streams.PipeTest.TestCloseReadSide` fails, but if you run that test by itself the test passes. I'm not quite sure of the reason why, though the fact that it consistently passes when it is run by itself might indicate that there is just some difference between the way XUnit runs its tests and the way MSTest works that is tripping up the test where it attempts to measure memory usage.

madelson commented:
@kal I was trying to get this to run locally but the .NET Core projects won't load for me. It says "The imported project "C:\Program Files (x86)\MSBuild\Microsoft\VisualStudio\v14.0\DotNet\Microsoft.DotNet.Props" was not found."". Is there something separate I need to install to build these projects? Is that external dependency required to get this working?

kal commented:
@madelson - you may need to upgrade .NET Core Tools on your machine. See https://github.com/NuGet/Home/issues/2616

madelson commented:
@kal ok everything is building now after installing https://www.microsoft.com/net/core#windows. However, I'm not sure how to get both sets of tests to run. I can still do Test menu->Run->All Tests, but that just runs each test once, presumably from the old test project. How did you go about explicitly running the core or .NET framework tests?

Another thing I noticed is that I have to manually force a build of the .NET Core projects; just changing a file does not seem to trigger this. Have you noticed this behavior?

Also, I figured out why that one test is failing in batch mode: it's because XUnit runs tests in parallel by default. I've disabled this via an assembly attribute and now the test works fine.

Commit:
Added a build script using Cake

http://cakebuild.net/

Commit:
Merge remote-tracking branch 'refs/remotes/madelson/master'

# Conflicts:
#	MedallionShell.Tests/GeneralTest.cs

kal commented:
@madelson Good catch with the XUnit tests! There doesn't seem to be any good way to get both sets of unit tests to run in Visual Studio - I guess it is because they use two different versions of the XUnit test runner. Perhaps that situation will improve once the next release of XUnit comes out (as the .NET Core build currently requires a pre-release version of XUnit). 

However, I've found that using Cake (http://cakebuild.net/) is a great way to automate all this stuff. I've added a commit with a simple Cake build file. To run the build you just run the build.ps1 in PowerShell - the script takes care of loading tools and dependencies automatically and the Default target will build and test both the .NET 4.5 and the .NET Core builds. The build file is the build.cake file - Cake uses C# so it should be pretty readable. To run a specific target you just add a -Target parameter to the script e.g. `build.ps1 -Target TestNet45`. 

I've also merged in your changes on master so we are conflict-free again :)

madelson commented:
@kal so when I try to run the cake build file I get an error:

```
An error occured when executing task 'CompileNetCore'.
Error: .NET Core CLI: Could not locate executable.
```

Any ideas? I downloaded and installed the .NET Core SDK for windows from here (https://www.microsoft.com/net/core#windows) but I still get the error. As we add these new dependencies I want to keep track of and document them so that anyone who downloads the source can get up and running.

kal commented:
@madelson - that's not something I've come across before, although I think on all my machines I have had .NET Core SDK already installed. I think the only requirement is that the dotnet command should be on the PATH (which should have happened when you installed the SDK). Can you open a Powershell terminal and just run the command `dotnet` ? Or even just do  `get-command dotnet`

madelson commented:
@kal Thanks. This helped me track down the issue. I generally use git bash for development so I had been invoking cake via:

```
powershell ./build.ps1
```

I figured that since I was starting ps each time I wouldn't need to restart git bash after installing the .NET Core SDK. Restarting git bash did the trick.

madelson commented on MedallionShell.NetCore/project.json:
```diff
@@ -0,0 +1,22 @@
+﻿{
+  "title": "MedallionShell",
+  "version": "1.0.0-*",
```
@kal why is this version specified here? Does it matter that it doesn't match the version from AssemblyInfo?

kal replied:
The version specified there would become the NuGet package version if you just generated the package directly from the project. Although during the build the values in AssemblyInfo.cs get used for generating the assembly attributes I actually think it would be a good idea to keep this version number in sync with your NuGet package version number.

madelson commented:
@kal at long last version 1.2.1 is now published which contains the new .NET core support. Thanks for all your help in pushing this through and your patience with my complete lack of knowledge regarding .NET core and multi-targeting in general.
